src/main/scala/is/hail/io/vcf/LoadVCF.scala

```diff
@@ -9,14 +9,187 @@ import is.hail.utils._
+          return true
+      }
+      i += 1
+    }
+    false
+  }
+
+  def symbolicAlt(alt: String): Boolean = {
+    assert(alt.length > 0)
+    val f = alt(0)
+    val l = alt(alt.length - 1)
+    if (f == '.' || l == '.' || (f == '<' && l == '>'))
+      return true
+
+    var i = 0
+    while (i < alt.length) {
+      val c = alt(i)
+      if (c == '[' || c == ']')
+        return true
+      i += 1
+    }
+    false
+  }
+}
+
+final class VCFLine(line: String) {
+  var pos: Int = 0
+
+  val abs = new ArrayBuilder[String]
+
+  def parseString(): String = {
+    val start = pos
+    while (pos < line.length && line(pos) != '\t')
+      pos += 1
+    val end = pos
+    line.substring(start, end)
+  }
+
+  def parseInt(): Int = {
+    var c: Char = 0
+
+    if (!(pos < line.length && line(pos) != '\t'))
+      fatal("empty integer field")
+    var v = VCFLine.numericValue(line(pos))
+    pos += 1
+    while (pos < line.length && line(pos) != '\t') {
+      v = v * 10 + VCFLine.numericValue(line(pos))
+      pos += 1
+    }
+    v
+  }
+
+  def skipField(): Unit = {
+    while (pos < line.length && line(pos) != '\t')
+      pos += 1
+  }
+
+  def parseStringInArray(): String = {
+    val start = pos
+    while (pos < line.length && line(pos) != '\t' && line(pos) != ',')
+      pos += 1
+    val end = pos
+    line.substring(start, end)
+  }
+
+  // leaves result in abs
+  def parseAltAlleles(): Unit = {
+    assert(abs.size == 0)
+
+    // . means no alternate alleles
+    if (pos < line.length
+      && line(pos) == '.'
+      && (pos + 1 == line.length || line(pos + 1) == '\t'))
+      return
+
+    abs += parseStringInArray()
+    while (pos < line.length && line(pos) != '\t') {
+      pos += 1 // comma
+      abs += parseStringInArray()
+    }
+  }
+
+  def nextField(): Unit = {
+    if (pos == line.length)
+      fatal("unexpected end of line")
+    pos += 1 // tab
+  }
+
+  // return false if it should be filtered
+  def parseAddVariant(rvb: RegionValueBuilder): Boolean = {
+    assert(pos == 0)
+
+    if (line.isEmpty || line(0) == '#')
+      return false
+
+    // CHROM (contig)
+    val contig = parseString()
+    nextField()
+
+    // POS (start)
+    val start = parseInt()
+    nextField()
+
+    skipField() // ID
+    nextField()
+
+    // REF
+    val ref = parseString()
+    if (VCFLine.invalidRef(ref)) {
+      warn(s"skipping variant with invalid REF field: $ref")
+      return false
+    }
+    nextField()
+
+    // ALT
+    parseAltAlleles()
+
+    var i = 0
+    while (i < abs.length) {
+      val alt = abs(i)
+      if (alt.isEmpty)
+        fatal("empty alternate allele")
+      if (VCFLine.symbolicAlt(alt)) {
+        warn(s"skipping variant with symbolic alternate allele in ALT field: $alt")
+        return false
+      }
+      i += 1
+    }
+
+    rvb.startStruct() // pk: Locus
+    rvb.addString(contig)
+    rvb.addInt(start)
+    rvb.endStruct()
+
+    rvb.startStruct() // v: Variant
+    rvb.addString(contig)
```

Comment:
We should think about shared pointers in the future.

Reply:
Shared pointers?  Tell me (or dev) more.

Reply:
In this case, we add the contig and ref strings in a bunch of places. The same string is written to the RV  several times, even though just copying the offset would work given current infrastructure (it would get duplicated on encoding, though).

Reply:
This is often called [String interning](https://en.wikipedia.org/wiki/String_interning). It seems like we wouldn't get a big win when ref and alt are only one character long. I think ref/alt might be better attacked with a representation that just uses 4-bits each.